Sources/SwiftFusion/Core/Tuple+EuclideanVectorN.swift

```diff
@@ -0,0 +1,159 @@
+  public static func + (_ lhs: Self, _ rhs: Self) -> Self {
+    return Tuple(head: lhs.head + rhs.head, tail: lhs.tail + rhs.tail)
+  }
+  public static func - (_ lhs: Self, _ rhs: Self) -> Self {
+    return Tuple(head: lhs.head - rhs.head, tail: lhs.tail - rhs.tail)
+  }
+  public static var zero: Self {
+    return Tuple(head: Head.zero, tail: Tail.zero)
+  }
+}
+
+extension Empty: Differentiable {
+  public typealias TangentVector = Self
+  public mutating func move(along direction: TangentVector) {}
+}
+
+extension Tuple: Differentiable
+where Head: Differentiable, Tail: Differentiable, Tail.TangentVector: TupleProtocol {
+  public typealias TangentVector = Tuple<Head.TangentVector, Tail.TangentVector>
+  public mutating func move(along direction: TangentVector) {
+    head.move(along: direction.head)
+    tail.move(along: direction.tail)
+  }
+}
+
+extension Empty: EuclideanVector {
+  @differentiable
+  public static func += (_ lhs: inout Self, _ rhs: Self) {}
+
+  @differentiable
+  public static func -= (_ lhs: inout Self, _ rhs: Self) {}
+
+  // MARK: - Scalar multiplication.
+
+  @differentiable
+  public static func *= (_ lhs: inout Self, _ rhs: Double) {}
+
+  // MARK: - Euclidean structure.
+
+  /// The inner product of `self` with `other`.
+  @differentiable
+  public func dot(_ other: Self) -> Double { return 0 }
+
+  // MARK: - Conversion to/from a concrete dynamically-sized `Vector`.
+
+  /// Creates an instance with the same elements as `vector`.
+  @differentiable
+  public init(_ vector: Vector) { self.init() }
+
+  /// A `Vector` containing the same elements as `self`.
+  @differentiable
+  public var vector: Vector { return Vector([]) }
+}
+
+extension Empty: EuclideanVectorN {
+  /// The dimension of the vector.
+  public static var dimension: Int { return 0 }
+
+  /// A standard basis of vectors.
+  public static var standardBasis: [Self] { return [] }
+}
+
+extension Tuple: EuclideanVector, EuclideanVectorN
+where Head: EuclideanVectorN, Tail: EuclideanVectorN {
+  @differentiable
+  public static func += (_ lhs: inout Self, _ rhs: Self) {
+    lhs.head += rhs.head
+    lhs.tail += rhs.tail
+  }
+
+  @differentiable
+  public static func -= (_ lhs: inout Self, _ rhs: Self) {
+    lhs.head -= rhs.head
+    lhs.tail -= rhs.tail
+  }
+
+  // MARK: - Scalar multiplication.
+
+  @differentiable
+  public static func *= (_ lhs: inout Self, _ rhs: Double) {
+    lhs.head *= rhs
+    lhs.tail *= rhs
+  }
+
+  // MARK: - Euclidean structure.
+
+  /// The inner product of `self` with `other`.
+  @differentiable
+  public func dot(_ other: Self) -> Double {
+    return head.dot(other.head) + tail.dot(other.tail)
+  }
+
+  // MARK: - Conversion to/from a concrete dynamically-sized `Vector`.
+
+  /// Creates an instance with the same elements as `vector`.
+  @differentiable
+  public init(_ vector: Vector) {
+    self.init(
+      head: Head(Vector(Array(vector.scalars[0..<Head.dimension]))),
```

Comment:
I have improved this. Hopefully it should be clearer now. It did require me to make some changes to `EuclideanVectorN`, but it's worth it because I'm going to need these changes in an upcoming PR.